presto-main/src/main/java/com/facebook/presto/server/remotetask/WorkerTaskStatusFetcher.java

```diff
@@ -0,0 +1,283 @@
+        if (!errorRateLimit.isDone()) {
+            errorRateLimit.addListener(this::scheduleNextRequest, executor);
+            return;
+        }
+
+        Request request = setContentTypeHeaders(isBinaryTransportEnabled, prepareGet())
+                .setUri(uriBuilderFrom(taskStatus.getSelf()).appendPath("status").build())
+                .setHeader(PRESTO_CURRENT_STATE, taskStatus.getState().toString())
+                .setHeader(PRESTO_MAX_WAIT, refreshMaxWait.toString())
+                .build();
+
+        ResponseHandler responseHandler;
+        if (isBinaryTransportEnabled) {
+            ArrayList<SmileCodec<TaskStatus>> toPass = new ArrayList<>();
+            toPass.add((SmileCodec<TaskStatus>) taskStatusCodec);
+
+            responseHandler = createFullSmileResponseHandler((SmileCodec<List<TaskStatus>>) toPass);
+        }
+        else {
+            responseHandler = createAdaptingJsonResponseHandler(unwrapJsonCodec(taskStatusCodec));
+        }
+
+        errorTracker.startRequest();
+        future = httpClient.executeAsync(request, responseHandler);
+        currentRequestStartNanos.set(System.nanoTime());
+        Futures.addCallback(future, new SimpleHttpResponseHandler<>(this, request.getUri(), stats), executor);
+    }
+
+    TaskStatus getTaskStatus()
+    {
+        return taskStatus.get();
+    }
+
+    @Override
+    public void success(TaskStatus value)
+    {
+        try (SetThreadName ignored = new SetThreadName("ContinuousTaskStatusFetcher-%s", taskId)) {
+            updateStats(currentRequestStartNanos.get());
+            try {
+                updateTaskStatus(value);
+                errorTracker.requestSucceeded();
+            }
+            finally {
+                scheduleNextRequest();
+            }
+        }
+    }
+
+    @Override
+    public void failed(Throwable cause)
+    {
+        try (SetThreadName ignored = new SetThreadName("ContinuousTaskStatusFetcher-%s", taskId)) {
+            updateStats(currentRequestStartNanos.get());
+            try {
+                // if task not already done, record error
+                TaskStatus taskStatus = getTaskStatus();
+                if (!taskStatus.getState().isDone()) {
+                    errorTracker.requestFailed(cause);
+                }
+            }
+            catch (Error e) {
+                onFail.accept(e);
+                throw e;
+            }
+            catch (RuntimeException e) {
+                onFail.accept(e);
+            }
+            finally {
+                scheduleNextRequest();
+            }
+        }
+    }
+
+    @Override
+    public void fatal(Throwable cause)
+    {
+        try (SetThreadName ignored = new SetThreadName("ContinuousTaskStatusFetcher-%s", taskId)) {
+            updateStats(currentRequestStartNanos.get());
+            onFail.accept(cause);
+        }
+    }
+
+    void updateTaskStatus(TaskStatus newValue)
```

Comment:
this method will take input list of tasks and it's status. May be a map of <Task Id, TaskStatus>